discourse_rock/rockcraft.yaml

```diff
@@ -4,46 +4,24 @@
 name: discourse
 summary: Discourse rock
 description: Discourse OCI image for the Discourse charm
-base: ubuntu:20.04
+base: ubuntu@22.04
 run-user: _daemon_
 license: Apache-2.0
 version: "1.0"
 platforms:
   amd64:
 parts:
-  install-node:
-    # We need nodejs so that yarn works correctly
-    # This allows us to control the version of node
+  install-node-tools:
+    # Install svgo, yarn and terser via npm by installing nodejs
+    # Installing per https://github.com/discourse/install-rails/blob/main/linux
     plugin: nil
     source: .
-    build-environment:
-      - NODE_VERSION: "18.18.2"
-      - ARCH: "x64"
     overlay-script: |
-      node_uri="https://nodejs.org/dist/v${NODE_VERSION}/node-v${NODE_VERSION}-linux-${ARCH}.tar.gz"
-      curl -Ls $node_uri | tar xzf - -C $CRAFT_OVERLAY/ --skip-old-files --no-same-owner --strip-components=1
-  install-yarn:
-    # We need yarn to compile discourse's assets
-    plugin: nil
-    after:
-      - install-node
-    source: .
-    build-environment:
-      - YARN_VERSION: "1.22.19"
-      - ARCH: "x64"
-    overlay-script: |
-      yarn_uri="https://github.com/yarnpkg/yarn/releases/download/v${YARN_VERSION}/yarn-v${YARN_VERSION}.tar.gz"
-      curl -Ls $yarn_uri | tar xzf - -C $CRAFT_OVERLAY/ --skip-old-files --no-same-owner --strip-components=1
-  install-terser:
-    # Terser is needed to compile discourse's assets
-    # We don't use the system package to avoid pulling ubuntu's nodejs package
-    # Related to: https://git.launchpad.net/~weii-wang/charm-k8s-discourse/commit/?h=fix-terser-precompile&id=ada4ba97b28ab53dd0a619b3a9325b7a981afac1
-    plugin: nil
-    source: .
-    after:
-      - install-node
-    overlay-script: |
-      $CRAFT_OVERLAY/usr/bin/node $CRAFT_OVERLAY/usr/bin/npm --prefix $CRAFT_OVERLAY/usr/local/ i -g terser
+      curl -sL https://deb.nodesource.com/setup_18.x | bash -
```

Comment:
We're loosing a lot of control here: we can't specify the exact version of node nor the one of yarn, we can't verify the shasum of the binaries.
That said, I understand that the previous way was not "source first". Should we change that to rebuild node in a future version of the rockfile ?

Reply:
That's a valid concern, however - you should be able to specify the version of the packages in apt and in npm, it is just that we're not doing it currently. I think we should aim for apt/snap first and then source, but also I'm in favor of following the install guidelines of the upstream projects as close as possible.

Reply:
@nrobinaubertin the refactored version now installs the tooling in a versioned manner like the original did, so that should be fine now imo. wdyt?